File: app/src/main/java/it/cammino/risuscito/GeneralIndex.java
package it.cammino.risuscito;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import it.cammino.risuscito.ui.ThemeableActivity;
import it.cammino.risuscito.utils.ThemeUtils;

public class GeneralIndex extends Fragment {

  private static final String PAGE_VIEWED = "pageViewed";
  final String TAG = getClass().getCanonicalName();
  @BindView(R.id.view_pager)
  ViewPager mViewPager;
  private MainActivity mMainActivity;
  private Unbinder mUnbinder;

  @Override
  public View onCreateView(
      @NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
    final View rootView = inflater.inflate(R.layout.tabs_layout, container, false);
    mUnbinder = ButterKnife.bind(this, rootView);

    mMainActivity = (MainActivity) getActivity();
    mMainActivity.setupToolbarTitle(R.string.title_activity_general_index);

    mViewPager.setAdapter(new SectionsPagerAdapter(getChildFragmentManager()));

    final TabLayout tabs = mMainActivity.mTabLayout;
    tabs.setVisibility(View.VISIBLE);
    mMainActivity.enableFab(false);
    if (!mMainActivity.isOnTablet()) mMainActivity.enableBottombar(false);
    if (savedInstanceState == null) {
      SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(getActivity());
      mViewPager.setCurrentItem(Integer.parseInt(pref.getString(Utility.DEFAULT_INDEX, "0")));
    } else mViewPager.setCurrentItem(savedInstanceState.getInt(PAGE_VIEWED, 0));
    if (!mMainActivity.isOnTablet()) tabs.setBackgroundColor(getThemeUtils().primaryColor());
    tabs.setupWithViewPager(mViewPager);

    return rootView;
  }

  @Override
  public void onDestroyView() {
    super.onDestroyView();
    mUnbinder.unbind();
  }

  @Override
  public void onSaveInstanceState(@NonNull Bundle outState) {
    super.onSaveInstanceState(outState);
    outState.putInt(PAGE_VIEWED, mViewPager.getCurrentItem());
  }

  private ThemeUtils getThemeUtils() {
    return mMainActivity.getThemeUtils();
  }

  private class SectionsPagerAdapter extends FragmentStatePagerAdapter {

    SectionsPagerAdapter(FragmentManager fm) {
      super(fm);
    }

    @Override
    public Fragment getItem(int position) {
      switch (position) {
        case 0:
          return new AlphabeticSectionFragment();
        case 1:
          return new NumericSectionFragment();
        case 2:
          return new ArgumentsSectionFragment();
        case 3:
          return new SalmiSectionFragment();
        case 4:
          return new IndiceLiturgicoFragment();
        default:
          return new AlphabeticSectionFragment();
      }
    }

    @Override
    public int getCount() {
      return 5;
    }

    @Override
    public CharSequence getPageTitle(int position) {
      //            Locale l = getActivity().getResources().getConfiguration().locale;
      Locale l =
          ThemeableActivity.getSystemLocalWrapper(getActivity().getResources().getConfiguration());
      switch (position) {
        case 0:
          return getString(R.string.letter_order_text).toUpperCase(l);
        case 1:
          return getString(R.string.page_order_text).toUpperCase(l);
        case 2:
          return getString(R.string.arg_search_text).toUpperCase(l);
        case 3:
          return getString(R.string.salmi_musica_index).toUpperCase(l);
        case 4:
          return getString(R.string.indice_liturgico_index).toUpperCase(l);
      }
      return null;
    }
  }
}

File: app/src/main/java/it/cammino/risuscito/GeneralIndex.kt
package it.cammino.risuscito

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.preference.PreferenceManager
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import com.google.android.material.tabs.TabLayoutMediator
import it.cammino.risuscito.databinding.TabsLayoutBinding
import it.cammino.risuscito.ui.LocaleManager.Companion.getSystemLocale
import it.cammino.risuscito.viewmodels.GeneralIndexViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class GeneralIndex : Fragment() {

    private var mMainActivity: MainActivity? = null

    private val mViewModel: GeneralIndexViewModel by viewModels()

    private val mPageChange: ViewPager2.OnPageChangeCallback = object : ViewPager2.OnPageChangeCallback() {
        override fun onPageSelected(position: Int) {
            Log.d(TAG, "onPageSelected: $position")
            mViewModel.pageViewed = position
        }
    }

    private var _binding: TabsLayoutBinding? = null

    // This property is only valid between onCreateView and
    // onDestroyView.
    private val binding get() = _binding!!

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = TabsLayoutBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroyView() {
        binding.viewPager.unregisterOnPageChangeCallback(mPageChange)
        _binding = null
        super.onDestroyView()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        mMainActivity = activity as? MainActivity
        mMainActivity?.setupToolbarTitle(R.string.title_activity_general_index)
        mMainActivity?.setTabVisible(true)
        mMainActivity?.enableFab(false)
        mMainActivity?.enableBottombar(false)

        binding.viewPager.adapter = IndexTabsAdapter(this)
        mMainActivity?.getMaterialTabs()?.let {
            TabLayoutMediator(it, binding.viewPager) { tab, position ->
                val l = getSystemLocale(resources)
                tab.text = when (position) {
                    0 -> getString(R.string.letter_order_text).toUpperCase(l)
                    1 -> getString(R.string.page_order_text).toUpperCase(l)
                    2 -> getString(R.string.arg_search_text).toUpperCase(l)
                    3 -> getString(R.string.salmi_musica_index).toUpperCase(l)
                    4 -> getString(R.string.indice_liturgico_index).toUpperCase(l)
                    else -> getString(R.string.letter_order_text).toUpperCase(l)
                }
            }.attach()
        }
        binding.viewPager.registerOnPageChangeCallback(mPageChange)

        lifecycleScope.launch {
            delay(500)
            if (savedInstanceState == null) {
                val pref = PreferenceManager.getDefaultSharedPreferences(context)
                binding.viewPager.currentItem = Integer.parseInt(pref.getString(Utility.DEFAULT_INDEX, "0")
                        ?: "0")
            } else
                binding.viewPager.currentItem = mViewModel.pageViewed
        }

    }

    private class IndexTabsAdapter(fragment: Fragment) : FragmentStateAdapter(fragment) {
        override fun getItemCount(): Int = 5

        override fun createFragment(position: Int): Fragment =
                when (position) {
                    0 -> SimpleIndexFragment.newInstance(0)
                    1 -> SimpleIndexFragment.newInstance(1)
                    2 -> SectionedIndexFragment.newInstance(0)
                    3 -> SimpleIndexFragment.newInstance(2)
                    4 -> SectionedIndexFragment.newInstance(1)
                    else -> SimpleIndexFragment.newInstance(0)
                }
    }

    companion object {
        internal val TAG = GeneralIndex::class.java.canonicalName
    }

}
